tests/net_dynamic_hb.rs

```diff
@@ -0,0 +1,201 @@
+                        .public_key();
+                    let _ = net[node_id]
+                        .algorithm_mut()
+                        .input(Input::Change(Change::Add(*pivot_node_id, pk)))
+                        .expect("failed to send `Add` input");
+                }
+
+                ChangeState::Complete(Change::Add(pivot_node_id, _)) => {
+                    println!("Node {:?} done adding.", node_id);
+                    // Node added, ensure it has been removed first.
+                    if awaiting_removal.contains(&node_id) {
+                        panic!(
+                            "Node {:?} reported a success `Add({}, _)` before `Remove({})`",
+                            node_id, pivot_node_id, pivot_node_id
+                        );
+                    }
+                    awaiting_addition.remove(&node_id);
+                }
+                _ => {
+                    println!("Unhandled change: {:?}", change);
+                }
+            }
+        }
+
+        // Record whether or not we received some output.
+        let has_output = !step.output.is_empty();
+
+        // Find the node's input queue.
+        let queue: &mut Vec<_> = queues
+            .get_mut(&node_id)
+            .expect("queue for node disappeared");
+
+        // Examine potential algorithm output.
+        for batch in step.output.into_iter() {
+            println!(
+                "Received epoch {} batch on node {:?}.",
+                batch.epoch(),
+                node_id,
+            );
+
+            for tx in batch.iter() {
+                // Remove the confirmed contribution from the input queue.
+                let index = queue.iter().position(|v| v == tx);
+                if let Some(idx) = index {
+                    assert_eq!(queue.remove(idx), *tx);
+                }
+
+                // Add it to the set of received outputs.
+                if !net[node_id].is_faulty() {
+                    expected_outputs
+                        .get_mut(&node_id)
+                        .expect("output set disappeared")
+                        .remove(tx);
+                }
+            }
+        }
+
+        // Check if we are done.
+        if expected_outputs.values().all(|s| s.is_empty())
+            && awaiting_addition.is_empty()
+            && awaiting_removal.is_empty()
+        {
+            // All outputs are empty and all nodes have removed and added the single pivot node.
+            break;
+        }
+
+        // If not done, check if we still want to propose something.
+        if has_output {
+            if !queue.is_empty() {
+                // Out of the remaining transactions, select a suitable amount.
+                let proposal = rand::seq::sample_slice(
+                    &mut rng,
+                    // FIXME: Use better numbers and proptest.
+                    queue.as_slice().subslice(0..10),
+                    10.min(3.min(queue.len())),
```

Comment:
Isn't that always equal to `3.min(queue.len())`? Or did you mean to write `total_txs` instead of `10`?
You could also flip the order to reduce the nesting:

```rust
queue.len().min(total_txs).min(3)
```
```